tensorflow_addons/losses/giou_loss.py

```diff
@@ -0,0 +1,140 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Implements GIoU loss."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+from tensorflow_addons.utils import keras_utils
+
+
+@keras_utils.register_keras_custom_object
+class GIoULoss(tf.keras.losses.Loss):
+    """Implements the GIoU loss function.
+
+    GIoU loss was first introduced in the Generalized Intersection over Union
+    paper (https://giou.stanford.edu/GIoU.pdf). GIoU is a enhance for model
+    which use IOU in object detection.
+
+    Usage:
+
+    ```python
+    gl = tfa.losses.GIoU()
+    boxes1 = tf.constant([[4.0, 3.0, 7.0, 5.0], [5.0, 6.0, 10.0, 7.0]])
+    boxes2 = tf.constant([[3.0, 4.0, 6.0, 8.0], [14.0, 14.0, 15.0, 15.0]])
+    loss = gl(boxes1,boxes2)
+    print('Loss: ', loss.numpy())  # Loss: [1.07500000298023224, 1.9333333373069763]
+    ```
+    Usage with tf.keras API:
+
+    ```python
+    model = tf.keras.Model(inputs, outputs)
+    model.compile('sgd', loss=tf.keras.losses.GIoULoss())
+    ```
+
+    Args
+      mode: one of ['giou', 'iou'], decided to calculate giou loss or iou loss.
+
+    Returns:
+      GIoU loss double `Tensor`.
+    """
+
+    def __init__(self,
+                 mode='giou',
+                 reduction=tf.keras.losses.Reduction.AUTO,
+                 name='giou_loss'):
+        super(GIoULoss, self).__init__(name=name, reduction=reduction)
```

Comment:
Would like to see some checks for `mode` here. Maybe raise `ValueError` if mode is neither `giou` nor `iou`.

Reply:
I think the assertion is still needed here as users could know if they do something wrong when they construct the metric.